### PR TITLE
Try to rewrite e2e tests using jest-puppeter preset

### LINKED ISSUE #8619: Create a "click on the block appender" test helper for e2e tests
See: 
https://github.com/WordPress/gutenberg/pull/8301#discussion_r207977720

Basically, we do `await page.click( '.editor-default-block-appender__content' );` in the e2e tests a lot, but really this could just be a test helper like `await clickBlockAppender( page )`. This change would:

1. State the intent/action more clearly
2. DRY up the code

### PR DESCRIPTION
**Pull request by gziolo**
## Description

Closes #8619.

Improves #6956. See known issues to find out what is still necessary to make it bulletproof.

Tries to fix intermittent e2e test failures. In particular an error with `preview` test:

> Starting to see a new one.
> 
> ```
> FAIL test/e2e/specs/preview.test.js
>   ● Preview › Should open a preview window for a new post
>     Protocol error (Runtime.callFunctionOn): Cannot find context with specified id undefined
>       
>       at Promise (node_modules/puppeteer/lib/Connection.js:198:56)
>       at CDPSession.send (node_modules/puppeteer/lib/Connection.js:197:12)
>       at ExecutionContext.evaluateHandle (node_modules/puppeteer/lib/ExecutionContext.js:71:75)
>       at ExecutionContext.evaluate (node_modules/puppeteer/lib/ExecutionContext.js:46:31)
>       at Frame.evaluate (node_modules/puppeteer/lib/FrameManager.js:299:20)
> ```
> 
> Example: #7841 https://travis-ci.org/WordPress/gutenberg/jobs/401939944

**This is try to use `jest-puppeteer` preset recommended in Jest docs:**

https://jestjs.io/docs/en/puppeteer#use-puppeteer-preset

The biggest difference is that we no longer use isolated browser instance per test suite but share one browser to speed up tests execution.

**It also enables some new methods to `expect` API.** See:
https://github.com/smooth-code/jest-puppeteer/tree/master/packages/expect-puppeteer#api

## How has this been tested?

`npm run test-e2e`

## Knows issues

We should open a follow up tasks for two issues.

### Preview opens new tab after publish

Steps to reproduce:
1. Type something in the title field.
1. Click the preview button.
1. Go back to the editor.
1. Publish post.
1. Click the preview button.
1. Go back to the editor.
1. Type something in the title field.
1. Click the preview button.
1. It shouldn't open a new tab. It should reuse the existing one.

### Too many undo levels

It happens only in the headless mode when running e2e tests. Sometimes it creates two undo levels for the first paragraph. When you modify test and add snapshots after each undo action you might see the following:

```js
exports[`undo Should undo to expected level intervals 5`] = `
“<!-- wp:paragraph -->
<p>This</p>
<!-- /wp:paragraph -->”
`;

exports[`undo Should undo to expected level intervals 6`] = `
“<!-- wp:paragraph -->
<p>Thi</p>
<!-- /wp:paragraph -->”
`;
```

This isn't expected behavior. It uncovers some edge case happening inside the logic which creates undo levels.

### CODE REVIEW

File: test/e2e/specs/preview.test.js [hunk 1]
```diff
@@ -148,19 +138,20 @@ describe( 'Preview', () => {
...
 		expect( previewPage.url() ).toBe( expectedPreviewURL );
 
 		// Return to editor to change title.
 		await editorPage.bringToFront();
 		await editorPage.type( '.editor-post-title__input', ' And more.' );
 
 		// Published preview should reuse same popup frame.
-		previewPage = await getOpenedPreviewPage();
+		// TODO: Fix code to reuse the same frame!
+		previewPage = await getOpenedPreviewPage( editorPage );
```
Comment (gziolo) [inline test/e2e/specs/preview.test.js]: This is the only place where we don't reuse the same frame. Very surprising given given that there is code comment which states it 😃

**Comment by gziolo** [comment]
Travis run for e2e tests is back to 7 minutes!

This is what I see locally:
```
Test Suites: 23 passed, 23 total
Tests:       3 skipped, 66 passed, 69 total
Snapshots:   15 passed, 15 total
Time:        93.87s, estimated 95s
Ran all test suites
```

File: test/e2e/specs/nux.test.js [hunk 2]
```diff
@@ -35,13 +32,7 @@ describe( 'New User Experience (NUX)', () => {
 	}
 
 	beforeEach( async () => {
-		await newDesktopBrowserPage();
-		await newPost( undefined, false );
-	} );
-
-	afterEach( async () => {
-		// Clear localStorage tips so they aren't persisted for the next test.
-		await clearLocalStorage();
```
Comment (youknowriad) [inline test/e2e/specs/nux.test.js]: Is this not needed anymore?

File: test/e2e/specs/multi-block-selection.test.js [hunk 3]
```diff
@@ -17,7 +15,7 @@ describe( 'Multi-block selection', () => {
 		const multiSelectedCssClass = 'is-multi-selected';
 
 		// Creating test blocks
-		await page.click( '.editor-default-block-appender' );
+		await page.click( '.editor-default-block-appender__content' );
```
Comment (youknowriad) [inline test/e2e/specs/multi-block-selection.test.js]: Just curious why this change?

File: test/e2e/specs/reusable-blocks.test.js [hunk 4]
```diff
@@ -1,27 +1,21 @@
...
-	newDesktopBrowserPage,
 	newPost,
 	pressWithModifier,
 	searchForBlock,
 } from '../support/utils';
 
 function waitForAndAcceptDialog() {
 	return new Promise( ( resolve ) => {
-		page.once( 'dialog', async ( dialog ) => {
-			await dialog.accept();
```
Comment (youknowriad) [inline test/e2e/specs/reusable-blocks.test.js]: Why are we removing this?

File: test/e2e/specs/sidebar-behaviour.test.js [hunk 5]
```diff
@@ -66,8 +61,7 @@ describe( 'Publishing', () => {
 
 	it( 'Should reopen sidebar the sidebar when resizing from mobile to desktop if the sidebar was closed automatically', async () => {
 		await setViewport( 'large' );
-		await newPost();
-		await setViewport( 'small' );
+		await newPost( { viewport: 'small' } );
```
Comment (youknowriad) [inline test/e2e/specs/sidebar-behaviour.test.js]: I think it's weird to have an option to set the viewport size as part of the `newPost` helper, I prefer a separate helper for that.

File: test/e2e/support/setup-test-framework.js [hunk 6]
```diff
@@ -1,23 +1,28 @@
...
 
 // The Jest timeout is increased because these tests are a bit slow
 jest.setTimeout( PUPPETEER_TIMEOUT || 100000 );
 
 beforeAll( async () => {
-	global.browser = await puppeteer.launch( {
-		headless: PUPPETEER_HEADLESS !== 'false',
-		slowMo: parseInt( PUPPETEER_SLOWMO, 10 ) || 0,
-	} );
+	enablePageDialogAccept();
```
Comment (youknowriad) [inline test/e2e/support/setup-test-framework.js]: I don't like that we're accepting dialogs globally, I prefer the explicitness we had before, because sometimes we want to test these dialogs

File: test/e2e/specs/multi-block-selection.test.js (hunk 3 above)
Comment (gziolo) [reply test/e2e/specs/multi-block-selection.test.js]: For consistency, we probably should wrap it with a utility method. Initially, I suspected that it might be the reason why undo test fails sometimes.

File: test/e2e/specs/reusable-blocks.test.js (hunk 4 above)
Comment (gziolo) [reply test/e2e/specs/reusable-blocks.test.js]: We have global handler so you don't need to accept it twice.

File: test/e2e/support/setup-test-framework.js (hunk 6 above)
Comment (gziolo) [reply test/e2e/support/setup-test-framework.js]: We need for the majority of test suites as far as I discovered today. There is `disablePageDialogAccept` method which would be useful in case when you want to handle it yourself. I'm happy to refactor, but then we would have to put it for every test which types something.

Comment (youknowriad) [reply test/e2e/support/setup-test-framework.js]: Ok 👍 It's fine

File: test/e2e/specs/sidebar-behaviour.test.js (hunk 5 above)
Comment (gziolo) [reply test/e2e/specs/sidebar-behaviour.test.js]: I can revert those changes. Previously, we were setting `large` viewport using `newDesktopBrowserPage` helper.

**Comment by youknowriad** [comment]
I'd appreciate if we extend our testing docs with steps to:

 - Run a unique test with a custom environment
 - Run a custom environment
 - Watch a unique test if possible?

Just trying to think of alternative workflows (other than CI) needed to debug and build tests.

File: test/e2e/specs/nux.test.js (hunk 2 above)
Comment (gziolo) [reply test/e2e/specs/nux.test.js]: We clear `localStorage` after every test suite. In addition, when opening the editor and `enableTips` flag is passed we clean all disabled tips.

**Comment by gziolo** [comment]
I brought back explicit calls for `setViewport` with 78825c0 as suggested by @youknowriad.

I  also added a change which ensures `npm run test-e2e` is executed with the latest codebase with aecf753 as suggested by @notnownikki.

File: test/e2e/specs/sidebar-behaviour.test.js (hunk 5 above)
Comment (gziolo) [reply test/e2e/specs/sidebar-behaviour.test.js]: Reverted 👍

**Comment by gziolo** [comment]
@youknowriad I opened a follow-up issue for your comment about documentation - #8319.

**Comment by gziolo** [comment]
Added another small tweak which improves the trigger which wraps up after tests cleanup: 384cf5e.

File: test/e2e/specs/preview.test.js [hunk 7]
```diff
@@ -6,96 +6,45 @@ import { parse } from 'url';
...
-					resolve( lastPreviewPage );
-				}
-
-				lastPreviewPage.once( 'load', onLastPreviewPageLoaded );
-				eventHandlers.push( [ lastPreviewPage, 'load', onLastPreviewPageLoaded ] );
-			} ) );
-		}
+	function sleep( time ) {
+		return new Promise( ( resolve ) => setTimeout( resolve, time ) );
+	}
```
Comment (noisysocks) [inline test/e2e/specs/preview.test.js]: Puppeteer already provides this function in the form of [`page.waitFor`](https://github.com/GoogleChrome/puppeteer/blob/v1.4.0/docs/api.md#pagewaitforselectororfunctionortimeout-options-args).

File: test/e2e/specs/preview.test.js [hunk 8]
```diff
@@ -126,41 +75,47 @@ describe( 'Preview', () => {
...
+		previewPage = await openPreviewPage( editorPage );
 		expect( previewPage.url() ).toBe( expectedPreviewURL );
 
 		// Return to editor to change title.
 		await editorPage.bringToFront();
 		await editorPage.type( '.editor-post-title__input', ' And more.' );
 
 		// Published preview should reuse same popup frame.
-		previewPage = await getOpenedPreviewPage();
+		// TODO: Fix an existing bug which opens a new tab.
```
Comment (noisysocks) [inline test/e2e/specs/preview.test.js]: Can we create an issue to track this work?

File: test/e2e/specs/reusable-blocks.test.js (hunk 4 above)
Comment (noisysocks) [reply test/e2e/specs/reusable-blocks.test.js]: Can we remove the entire `waitForAndAcceptDialog` function, then?

**Review by noisysocks — COMMENTED** [review]
Left a few minor comments. Would love to get this in very soon—it's hard to review PRs when they're all 🔴! 😅

**Comment by noisysocks** [comment]
Heads up that https://github.com/WordPress/gutenberg/pull/8330 contains a fix for the `preview` E2E test as well.

File: test/e2e/specs/preview.test.js (hunk 8 above)
Comment (gziolo) [reply test/e2e/specs/preview.test.js]: #8330 - thanks :)
Issue is filed: #8318.

File: test/e2e/specs/reusable-blocks.test.js (hunk 4 above)
Comment (gziolo) [reply test/e2e/specs/reusable-blocks.test.js]: We use this `Promise`, so not quite sure 🤷‍♂️

File: test/e2e/specs/preview.test.js (hunk 7 above)
Comment (youknowriad) [reply test/e2e/specs/preview.test.js]: Ok, that's fine, can we add an inline comment with this explanation?

Comment (notnownikki) [reply test/e2e/specs/preview.test.js]: Sure, in fact I've already changed this code locally and used the editor page to call `waitFor`, with an inline comment :)

**Comment by notnownikki** [comment]
Latest code has been run locally 20 times in a row with no failures. It slightly changes the fix in master to ensure the navigation and click promises resolve in a fixed order, and adds a step which closes the preview page partway through the test, because sometimes it was losing the window name and causing the test to hang (because it had two preview pages open).

File: test/e2e/specs/multi-block-selection.test.js (hunk 3 above)
Comment (aduth) [reply test/e2e/specs/multi-block-selection.test.js]: This updated class is the one which has the events bound, so it's more "correct".

File: test/e2e/specs/preview.test.js [hunk 9]
```diff
@@ -6,23 +6,36 @@ import { parse } from 'url';
...
-		await newDesktopBrowserPage();
-	} );
-
 	beforeEach( async () => {
 		await newPost();
 	} );
 
+	async function openPreviewPage( editorPage ) {
+		let openTabs = await browser.pages();
+		expect( openTabs ).toHaveLength( 2 );
```
Comment (aduth) [inline test/e2e/specs/preview.test.js]: This function seems difficult to maintain, particularly if we add any more test cases, or "resolve" the issue where we have a blank open tab present in the test browser.

Do we really care if there are two tabs?
Couldn't we express the "wait for" as `+ 1` of however many open tabs there are initially?

File: test/e2e/specs/preview.test.js [hunk 10]
```diff
@@ -6,23 +6,36 @@ import { parse } from 'url';
...
+		await editorPage.click( '.editor-post-preview' );
+
+		// Wait for the new tab to open.
+		while ( 3 !== openTabs.length ) {
+			await editorPage.waitFor( 250 );
```
Comment (aduth) [inline test/e2e/specs/preview.test.js]: I thought we might have had a lint rule forbidding this. Timeouts are generally discouraged. In this case, it's similar in purpose to [`waitForFunction`](https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#pagewaitforfunctionpagefunction-options-args) which is effectively implemented by polling by default, though using `requestAnimationFrame`. Similarly, I'd like to see a much smaller interval here, even as low as `10`, `1`, or `0`. `250` is too long.

File: test/e2e/specs/preview.test.js [hunk 11]
```diff
@@ -94,11 +99,14 @@ describe( 'Preview', () => {
...
 		await publishPost();
 		await Promise.all( [
-			page.waitForFunction( () => ! document.querySelector( '.editor-post-preview' ) ),
-			page.click( '.editor-post-publish-panel__header button' ),
+			editorPage.waitForFunction( () => ! document.querySelector( '.editor-post-preview' ) ),
+			editorPage.click( '.editor-post-publish-panel__header button' ),
 		] );
 		expectedPreviewURL = await editorPage.$eval( '.notice-success a', ( node ) => node.href );
-		await waitForPreviewNavigation( previewPage );
+		// At this point, the preview page can lose its name, so we close it and open a fresh one
```
Comment (aduth) [inline test/e2e/specs/preview.test.js]: What's meant by "lose its name" ? I don't know that this is relevant any longer after #8330 . There should only ever be one preview window.

File: test/e2e/support/setup-test-framework.js [hunk 12]
```diff
@@ -1,23 +1,30 @@
...
 
 // The Jest timeout is increased because these tests are a bit slow
 jest.setTimeout( PUPPETEER_TIMEOUT || 100000 );
 
 beforeAll( async () => {
-	global.browser = await puppeteer.launch( {
-		headless: PUPPETEER_HEADLESS !== 'false',
-		slowMo: parseInt( PUPPETEER_SLOWMO, 10 ) || 0,
-	} );
+	await setViewport( 'large' );
```
Comment (aduth) [inline test/e2e/support/setup-test-framework.js]: Should we set this `beforeEach` ? Seems like it might otherwise be prone to leakage between tests (particularly looking at `sidebar-behaviour.test.js`).

File: test/e2e/support/setup-test-framework.js [hunk 13]
```diff
@@ -55,8 +52,11 @@ afterAll( async () => {
...
 		await page.click( '#doaction' );
-		await page.waitForNavigation();
+		await page.waitForXPath(
+			'//*[contains(@class, "updated notice")]/p[contains(text(), "moved to the Trash.")]'
+		);
 	}
 
-	await browser.close();
+	disablePageDialogAccept();
+	await clearLocalStorage();
```
Comment (aduth) [inline test/e2e/support/setup-test-framework.js]: This seems like something we ought to do before tests start, since `afterAll` might not be run if the test process is terminated abruptly, and would thus leave the browser in a tampered state for the next run?

File: test/e2e/support/utils.js [hunk 14]
```diff
@@ -86,31 +86,22 @@ export async function visitAdmin( adminPath, query ) {
 	}
 }
 
-export async function newPost( postType, disableTips = true ) {
+export async function newPost( { postType, enableTips = false } = {} ) {
 	await visitAdmin( 'post-new.php', postType ? 'post_type=' + postType : '' );
 
-	if ( disableTips ) {
+	const tipsEnabled = await page.evaluate( () => wp.data.select( 'core/nux' ).areTipsEnabled() );
+	if ( tipsEnabled && ! enableTips ) {
```
Comment (aduth) [inline test/e2e/support/utils.js]: It feels like we should be able to express this function more simply. Why do we need to check `areTipsEnabled` at all? Shouldn't this be accounted for by the effect of the `dispatch` ?

```js
await page.evaluate( ( _enableTips ) => {
	const action = _enableTips ? 'enableTips' : 'disableTips';
	wp.data.dispatch( 'core/nux' )[ action ]();
}, enableTips );

if ( enableTips ) {
	await page.reload();
}
```

File: test/e2e/support/utils.js [hunk 15]
```diff
@@ -292,3 +283,23 @@ export async function pressTimes( key, count ) {
 export async function clearLocalStorage() {
 	await page.evaluate( () => window.localStorage.clear() );
 }
+
+async function acceptPageDialog( dialog ) {
```
Comment (aduth) [inline test/e2e/support/utils.js]: JSDoc.

File: test/e2e/specs/preview.test.js (hunk 11 above)
Comment (notnownikki) [reply test/e2e/specs/preview.test.js]: At this point in the test, the existing preview window's `window.name` can be blank, and so a second one opens. I ran the test in non-headless mode and double checked. It might not happen always, but it happened enough for me to have to do this to get the test to pass consistently.

Comment (aduth) [reply test/e2e/specs/preview.test.js]: I would be more curious in understanding why the preview name becomes blank, as that sounds like a bug in and of itself.

Comment (notnownikki) [reply test/e2e/specs/preview.test.js]: Oh, agreed totally :) Would you be ok if we open an issue and addressed it in another PR? The e2e intermittent test failures have blocked us recently. I'd be happy to start investigating immediately.

Comment (notnownikki) [reply test/e2e/specs/preview.test.js]: For me, the tests fail sometimes without this diff, and pass every time with it.

I'll open an issue and look at it.

Comment (notnownikki) [reply test/e2e/specs/preview.test.js]: I've opened an issue here https://github.com/WordPress/gutenberg/issues/8367

Comment (notnownikki) [reply test/e2e/specs/preview.test.js]: Also, updated the comment.

File: test/e2e/support/setup-test-framework.js (hunk 12 above)
Comment (gziolo) [reply test/e2e/support/setup-test-framework.js]: Yes, we can.

File: test/e2e/specs/preview.test.js (hunk 10 above)
Comment (gziolo) [reply test/e2e/specs/preview.test.js]: `waitForFunction` doesn't seem to be a good fit because it is evaluated in browser context.

Comment (notnownikki) [reply test/e2e/specs/preview.test.js]: @aduth With `targetcreated` not being reliable, and the cause of most of the intermittent failures I've run up against in travis - the "Execution context lost" failure message was caused by that. So I don't see any other alternative than a timeout. Happy to change it to 10 though.

https://github.com/WordPress/gutenberg/pull/8301#discussion_r206726817 is where it was agreed to use `waitFor`. But I'm happy to change it back to a custom `sleep` function again if that's preferred.

File: test/e2e/support/setup-test-framework.js (hunk 13 above)
Comment (gziolo) [reply test/e2e/support/setup-test-framework.js]: Let's try to move it to `beforeEach` to ensure all tests start with the clean state. We can move to `beforeAll` if that turns out to be too aggressive.

File: test/e2e/support/utils.js (hunk 14 above)
Comment (gziolo) [reply test/e2e/support/utils.js]: Yeah nice idea 👍

File: test/e2e/specs/preview.test.js (hunk 9 above)
Comment (gziolo) [reply test/e2e/specs/preview.test.js]: Yes, I had the same feeling when I saw it for the first time.

Comment (notnownikki) [reply test/e2e/specs/preview.test.js]: Yeah that's fair, I was trying to check here for the double preview tab issue, but as that's a very rare intermittent issue I'll change this and continue investigating it elsewhere.

**Comment by gziolo** [comment]
@aduth I addressed all your comments, it should be ready for another review.

File: test/e2e/support/setup-test-framework.js [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+jest.setTimeout( PUPPETEER_TIMEOUT || 100000 );
+
+// Before every test suite run, delete all content created by the test. This ensures
+// other posts/comments/etc. aren't dirtying tests and tests don't depend on
+// each other's side-effects.
+beforeAll( async () => {
+	enablePageDialogAccept();
+
+	// Visit `/wp-admin/edit.php` so we can see a list of posts and delete them.
+	await visitAdmin( 'edit.php' );
```
Comment (gziolo) [inline test/e2e/support/setup-test-framework.js]: @tofumatt - so we now do cleanup before every test suite starts.

**Comment by notnownikki** [comment]
I had to run the tests twice on The Machine of Doom to get these to pass, there was an intermittent fail on `test/e2e/specs/reusable-blocks.test.js` but `test/e2e/specs/preview.test.js` seems solid now. I've seen the reusable blocks test fail before in CI, but rarely, and a restart of the job usually passes.

**Comment by notnownikki** [comment]
In the past few CI runs on this PR, the following tests:

```
test/e2e/specs/reusable-blocks.test.js
test/e2e/specs/managing-links.test.js
test/e2e/specs/style-variation.test.js
```

intermittently failed. Those tests are moving the mouse, and then trying to click something that appears based on that mouse movement. I'm thinking we should be doing a `waitForSelector` before we try the click, so make sure we're not trying it too fast and trying to click something that hasn't appeared yet.

**Comment by gziolo** [comment]
I just restarted Travis build because `test/e2e/specs/style-variation.test.js` failed. Button not found, so it seems you are totally right. We should update those tests. Can you try if your solution helps?

**Comment by notnownikki** [comment]
@gziolo on it now :)

**Comment by notnownikki** [comment]
I've noticed that the more load my poor little laptop is under, the more frequently these intermittent failures appear, so with that in mind, https://github.com/WordPress/gutenberg/pull/8301/commits/31fe2be2cd7c6982008bfdc7ce07bd98109e265b does a few things. 1) It sends more mouse movement events so that chrome is less likely to miss things and not update. 2) It waits for the desired selector after mouse movements, so chrome has time to update and render. 3) It waits for a selector at the beginning of the undo test, which sometimes wasn't there right away when the machine running the test is under extreme load.

These pass now even when I'm having my laptop do disk benchmarks, show HD video, run chromium with many many tabs open, and play FTL at the same time.

File: test/e2e/support/setup-test-framework.js (hunk 16 above)
Comment (tofumatt) [reply test/e2e/support/setup-test-framework.js]: This seems good to me.

**Review by tofumatt — APPROVED** [review]
I ran these locally and they worked for me.

`npm run test-e2e:watch` has been flaky for me for the past while and seems to run better with this as well.

This is way nicer. 🚢

File: test/e2e/specs/change-detection.test.js [hunk 17]
```diff
@@ -144,7 +137,7 @@ describe( 'Change detection', () => {
 	} );
 
 	it( 'Should prompt if content added without save', async () => {
-		await page.click( '.editor-default-block-appender' );
+		await page.click( '.editor-default-block-appender__content' );
```
[the remark this answers is not in the document]
Comment (tofumatt) [reply test/e2e/specs/change-detection.test.js]: I filed #8619, but if you want to do it in this PR feel free 😉. If not though no worries.

File: test/e2e/specs/adding-blocks.test.js [hunk 18]
```diff
@@ -84,7 +81,8 @@ describe( 'adding blocks', () => {
 		// Using the between inserter
 		const insertionPoint = await page.$( '[data-type="core/quote"] .editor-block-list__insertion-point-button' );
 		const rect = await insertionPoint.boundingBox();
-		await page.mouse.move( rect.x + ( rect.width / 2 ), rect.y + ( rect.height / 2 ) );
+		await page.mouse.move( rect.x + ( rect.width / 2 ), rect.y + ( rect.height / 2 ), { steps: 10 } );
```
[the remark this answers is not in the document]
Comment (tofumatt) [reply test/e2e/specs/adding-blocks.test.js]: I saw this `{ steps: 10 }` argument across the tests though; if they need to stay *and* you don't want to add the eslint rule now, please file a follow-up issue for it as I think it's important 😄

[the remark this answers is not in the document]
Comment (aduth) [reply test/e2e/specs/adding-blocks.test.js]: I'll second the worry about `steps`. It's not entirely clear to me why they're necessary (aside from the obvious "it makes the tests green").

Is it a bug with Puppeteer? If so, we should file an issue.
Is it our application requiring multiple mousemove events to register that the effect is to take place? If so, we should know about this and consider whether it's a bug or intentional.

How also do we communicate to developers if and when they'll need to provide this in their tests? This may speak to @tofumatt 's point about an ESLint rule.

But I'd rather we first have a good understanding of why it's here.

[the remark this answers is not in the document]
Comment (tofumatt) [reply test/e2e/specs/adding-blocks.test.js]: Removing the `steps` call reliably caused many of my tests to fail though, so it seems they're needed. I just can't see why from the code.

**Comment by tofumatt** [comment]
I'm seeing:

```
 PASS  test/e2e/specs/preview.test.js (11.092s)
 PASS  test/e2e/specs/adding-blocks.test.js (7.049s)
 PASS  test/e2e/specs/publishing.test.js
 PASS  test/e2e/specs/hooks-api.test.js (7.705s)
 PASS  test/e2e/specs/nux.test.js (7.004s)
 PASS  test/e2e/specs/templates.test.js
 PASS  test/e2e/specs/plugins-api.test.js
 FAIL  test/e2e/specs/multi-block-selection.test.js
  ● Multi-block selection › Should select/unselect multiple blocks

    No node found for selector: .editor-default-block-appender__content

      at assert (node_modules/puppeteer/lib/helper.js:282:11)
      at Frame.click (node_modules/puppeteer/lib/FrameManager.js:590:5)
```

Failures intermittently, which is concerning. 😢

**Comment by notnownikki** [comment]
/me glares at the new intermittent failure.

@aduth the steps are there to make sure that an unresponsive / slowed Chrome responds to at least one of the events. These failures show up far more often when Travis is busy - they're rare in the mornings here in the UK, which is when we usually wait to get tests passing.

@tofumatt for the new interesting failures:

```      19 | 		// Should display the Post Formats UI.
      20 | 		const postFormatsUi = await page.$( '.editor-post-format' );
    > 21 | 		expect( postFormatsUi ).not.toBeNull();```

and

```    Error: failed to find element matching selector ".edit-post-post-status .my-post-status-info-plugin"
```

that's again most likely Chrome not responding quickly enough for the tests. Happens a lot when Travis is busy (that is, when America is awake and using it in earnest!)

https://github.com/WordPress/gutenberg/pull/8301#discussion_r207977720 - @gziolo tells me that there are helpers for this that `jest-puppeteer` provides. I'm not yet that familiar with that package though.

**Comment by tofumatt** [comment]
I've seen that failure (intermittently) locally too. 😢

**Comment by gziolo** [comment]
> PUPPETEER_HEADLESS=false PUPPETEER_SLOWMO=100 npm run test-e2e test/e2e/specs/templates.test.js test/e2e/specs/style-variation.test.js

This fails locally for me, when templates test suite is loaded first. Debugging.

File: test/e2e/specs/change-detection.test.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (gziolo) [reply test/e2e/specs/change-detection.test.js]: Addressed with 3bf1e49.

File: test/e2e/specs/preview.test.js [hunk 19]
```diff
@@ -31,34 +45,26 @@ describe( 'Preview', () => {
 	 *
 	 * @return {Promise} Promise resolving once navigation completes.
 	 */
-	function waitForPreviewNavigation( previewPage ) {
-		return Promise.all( [
-			previewPage.waitForNavigation(),
-			page.click( '.editor-post-preview' ),
-		] );
+	async function waitForPreviewNavigation( previewPage ) {
+		const previewNav = previewPage.waitForNavigation();
```
[the remark this answers is not in the document]
Comment (gziolo) [reply test/e2e/specs/preview.test.js]: Thanks 🙇

**Comment by gziolo** [comment]
It passed 2 times after recent tweaks. We should merge as soon as someone confirms I didn't break anything else 😅

**Comment by notnownikki** [comment]
:ship: 

Works well on my poor little laptop even when it's under load! This is a _huge_ improvement. Thank you!

**Comment by gziolo** [comment]
Thank you all for commits, reviews, and testing. Let's iterate on this as it is still not perfect 🚀
